server/src/main/java/org/opensearch/index/engine/Engine.java

```diff
@@ -1309,6 +1330,26 @@ public void failEngine(String reason, @Nullable Exception failure) {
         }
     }
 
+    /**
+     * Cleanup all unreferenced files generated during failed segment merge. This resets shard state to last Lucene
+     * commit.
+     */
+    private void cleanUpUnreferencedFiles() {
```

Comment:
We can have stats for cleanup during merge failures. But I think that it can be a separate change (can go in separate release), as we are logging whenever we are doing a cleanup. Will keep this thread open asking for others opinion as well.